common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/trajectory.hpp

```diff
@@ -341,6 +341,61 @@ double calcArcLength(const T & points)
 
   return calcSignedArcLength(points, 0, points.size() - 1);
 }
+
+/**
+ * Example Usage of calcSignedArcLength
+ * const size_t ego_seg_idx = findNearestSegmentIndexWithDistYawThreshold(points, ego_pose, ego_dist_threshold, ego_yaw_threshold);
+ * const size_t obj_seg_idx = findNearestSegmentIndexWithDistThreshold(points, obj_pos, obj_dist_threshold);
+ * const double signed_length = calcSignedArcLength(points, ego_pose.position, ego_seg_idx, obj_pos,
+ * obj_seg_idx);
+ */
+template <class T>
+double calcSignedArcLength(
+  const T & points, const geometry_msgs::msg::Point & src_point, const size_t src_seg_idx,
+  const geometry_msgs::msg::Point & dst_point, const size_t dst_seg_idx)
+{
+  validateNonEmpty(points);
+
+  const double signed_length_on_traj = calcSignedArcLength(points, src_seg_idx, dst_seg_idx);
+  const double signed_length_src_offset =
+    calcLongitudinalOffsetToSegment(points, src_seg_idx, src_point);
+  const double signed_length_dst_offset =
+    calcLongitudinalOffsetToSegment(points, dst_seg_idx, dst_point);
+
+  return signed_length_on_traj - signed_length_src_offset + signed_length_dst_offset;
+}
+
+template <class T>
+size_t findNearestSegmentIndexWithDistYawThreshold(
+  const T & points, const geometry_msgs::msg::Pose & pose, const double dist_threshold,
+  const double yaw_threshold)
+{
+  validateNonEmpty(points);
+
+  const auto nearest_with_dist_yaw_thresh = findNearestSegmentIndex(points, pose, dist_threshold, yaw_threshold);
+  if (nearest_with_dist_yaw_thresh) {
+    return nearest_with_dist_yaw_thresh.get();
+  }
+
+  return findNearestSegmentIndexWithDistThreshold(points, pose, dist_threshold);
+}
+
+template <class T>
+size_t findNearestSegmentIndexWithDistThreshold(
+  const T & points, const geometry_msgs::msg::Point & pos, const double dist_threshold)
+{
+  validateNonEmpty(points);
+
+  geometry_msgs::msg::Pose tmp_pose;
+  tmp_pose.position = pos;
+
+  const auto nearest_with_dist_thresh = findNearestSegmentIndex(points, tmp_pose, dist_threshold, std::numeric_limits<double>::max());
```

Comment:
I think it is better to use `std::optional` for yaw and dist threshold.
how about this?
```
findNearestSegmentIndex(points, tmp_pose, std::optional<double> dist_threshold, std::optional<double> yaw_threshold);
```

```suggestion
  const auto nearest_with_dist_thresh = findNearestSegmentIndex(points, tmp_pose, dist_threshold, std::nullopt);
```